dgl_sparse/src/sddmm.cc

```diff
@@ -0,0 +1,130 @@
+/**
+ *  Copyright (c) 2022 by Contributors
+ * @file sddmm.cc
+ * @brief DGL C++ sparse SDDMM operator implementation.
+ */
+// clang-format off
+#include <sparse/dgl_headers.h>
+// clang-format on
+
+#include <sparse/sparse_matrix.h>
+#include <sparse/spmm.h>
+#include <torch/script.h>
+
+#include "./utils.h"
+
+namespace dgl {
+namespace sparse {
+
+using namespace torch::autograd;
+
+class SDDMMAutoGrad : public Function<SDDMMAutoGrad> {
+ public:
+  static torch::Tensor forward(
+      AutogradContext* ctx, const c10::intrusive_ptr<SparseMatrix>& sparse_mat,
+      torch::Tensor mat1, torch::Tensor mat2_tr);
+
+  static tensor_list backward(AutogradContext* ctx, tensor_list grad_outputs);
+};
+
+void _SDDMMSanityCheck(
+    const c10::intrusive_ptr<SparseMatrix>& sparse_mat, torch::Tensor mat1,
+    torch::Tensor mat2) {
+  CHECK_EQ(mat1.dim(), 2)
+      << "SDDMM: the first dense matrix should be two-dimensional.";
+  CHECK_EQ(mat2.dim(), 2)
+      << "SDDMM: the second dense matrix should be two-dimensional.";
+  CHECK_EQ(sparse_mat->shape()[0], mat1.size(0))
+      << "SDDMM: the first dense matrix should have the same first dimension "
+         "as the sparse matrix";
+  CHECK_EQ(sparse_mat->shape()[1], mat2.size(1))
+      << "SDDMM: the seond dense matrix should have the same second dimension "
+         "as the sparse matrix";
+  CHECK_EQ(mat1.size(1), mat2.size(0))
+      << "SDDMM: the second dimension of the first dense matrix should be "
+         "equal to the first dimension of the second dense matrix.";
+  CHECK_EQ(mat1.dtype(), mat2.dtype())
+      << "SDDMM: the two dense matrices should have the same dtype.";
+}
+
+torch::Tensor SDDMMImpl(
+    const c10::intrusive_ptr<SparseMatrix>& sparse_mat, torch::Tensor mat1,
+    torch::Tensor mat2_tr) {
+  const int64_t out_row = sparse_mat->nnz();
+  const std::vector<int64_t> shape({out_row});
+  auto ret = torch::zeros(shape, mat1.options());
+  const std::string op = "dot";
+  auto dgl_mat1 = TorchTensorToDGLArray(mat1);
+  auto dgl_mat2_tr = TorchTensorToDGLArray(mat2_tr);
+  auto dgl_ret = TorchTensorToDGLArray(ret);
+  // Use CSR if the sparse matrix has CSR or does not have COO. Otherwise use
+  // COO.
+  if (sparse_mat->HasCSR() || !sparse_mat->HasCOO()) {
+    auto csr = CSRToOldDGLCSR(sparse_mat->CSRPtr());
+    aten::CSRSDDMM(
+        op.c_str(), csr, dgl_mat1, dgl_mat2_tr, dgl_ret, 0 /* Lhs target: u */,
+        2 /* rhs target: v */);
+  } else {  // COO
+    auto coo = COOToOldDGLCOO(sparse_mat->COOPtr());
+    aten::COOSDDMM(
+        op.c_str(), coo, dgl_mat1, dgl_mat2_tr, dgl_ret, 0 /* Lhs target: u */,
+        2 /* rhs target: v */);
+  }
+  return ret;
+}
+
+torch::Tensor SDDMMAutoGrad::forward(
+    AutogradContext* ctx, const c10::intrusive_ptr<SparseMatrix>& sparse_mat,
+    torch::Tensor mat1, torch::Tensor mat2) {
```

Comment:
If SDDMM is formulated as `(mat1 @ mat2) * sparse_mat`, then it should multiply the nonzero values of the sparse matrix too (and therefore need to support autograd of the nonzero values).

Reply:
From the doc https://docs.dgl.ai/en/latest/generated/dgl.mock_sparse.sddmm.html#dgl.mock_sparse.sddmm.
SDDMM is defined as (B @ C) * dglsp.ones_like(A) + A.

Reply:
That is not `SDDMM` but `sampled_addmm` following Pytorch (https://pytorch.org/docs/stable/generated/torch.sparse.sampled_addmm.html). I think the most common used `SDDMM` is `(B @ C) * dglsp.ones_like(A) * A` (http://tensor-compiler.org/docs/machine_learning.html.)

Reply:
> If SDDMM is formulated as (mat1 @ mat2) * sparse_mat, then it should multiply the nonzero values of the sparse matrix too (and therefore need to support autograd of the nonzero values).

The multiply is handled by libtorch.